### PR TITLE
Update for Neo4j 4.0

### LINKED ISSUE #486: Neo4j 4.0 neo4j:// support
The `bolt+routing://` has been replaced by `neo4j://` in Neo4j 4.0. This is currently not accepted.

### PR DESCRIPTION
**Pull request by JVemmer**
Hi @aanastasiou,

This is the combined pull request I mentioned in #490, #491 and #489.
It contains a bunch of changes, focused primarily on getting neomodel to play nice with Neo4j >= 4.0.
It also fixed a few tests that were no longer functional, or were not working in the first place (in any version of Neo4j I could find).
This was tested with the [supported versions of Neo4j](https://neo4j.com/developer/kb/neo4j-supported-versions/) (at the time of writing that is 3.4, 3.5 and 4.0), both via the test framework and in production. All tests passed.
This pull request fixes #486, #487, #488 and #489.
It may also have a significant impact on #485.

**Security consideration:** Neo4j 4.0 no longer adds a self-signed certificate, and therefore one of the commits here disables ENCRYPTED_CONNECTION. This has no meaningful effect on local instances (running via localhost), but can have significant security impact when submitting things via network. I would recommend at least updating the documentation to highlight this, and in the long term investigate whether we can do anything to alleviate this.
**Notes for people using this in production systems:** You can re-enable ENCRYPTED_CONNECTION as soon as you have created your own certificate, and ensured that it is trusted by your machine. So this is not just a full drop of encrypted connections, merely Neo4j no longer doing the work for you itself (discussions on self-signed certificates and security are out of the scope here).

### CODE REVIEW

**Comment by aanastasiou** [comment]
Thanks, just a quick note, it is best if you keep each change in its own PR as this makes merging and editing much easier.

**Comment by marcusze** [comment]
> Security consideration: Neo4j 4.0 no longer adds a self-signed certificate, and therefore one of the commits here disables ENCRYPTED_CONNECTION. 

In the official neo4j python driver I can connect with `GraphDatabase.driver("bolt://172.18.0.1:7687", auth=("neo4j", "123"), encrypted=False)` 
I would be great if the same would be possible in neomodel:
`db.set_connection('bolt://neo4j:neo4j@localhost:7687', encrypted=False)`

@aanastasiou, is there currently a way of enforcing non-encrypted db connections?

**Comment by marcusze** [comment]
`config.ENCRYPTED_CONNECTION = False`
Does the magic 👍

**Comment by adrianicv** [comment]
Hi,
is there any work left to do to merge this pull request?
With the limitation of not knowing deeply the code, I'm willing to help to push this forward and release a version compatible with neo4j v4.0
Thanks for the amazing job with this library!

**Comment by kvanvalkenhoef** [comment]
Nice work @JVemmer! Did these changes get accepted into the master branch? otherwise how could i make use of it anyway. want to be compatible with Neo4j 4.0

File: neomodel/config.py [hunk 1]
```diff
@@ -1,5 +1,5 @@
 AUTO_INSTALL_LABELS = False
 DATABASE_URL = 'bolt://neo4j:neo4j@localhost:7687'
 FORCE_TIMEZONE = False
-ENCRYPTED_CONNECTION = True
-MAX_POOL_SIZE = 50
+ENCRYPTED_CONNECTION = False
+MAX_POOL_SIZE = 50
```
Comment (mikekeda) [inline neomodel/config.py:5]: add a new line to the end of file

**Comment by JVemmer** [comment]
> Nice work @JVemmer! Did these changes get accepted into the master branch? otherwise how could i make use of it anyway. want to be compatible with Neo4j 4.0

It is currently not merged into the master branch. If you wish to use the changes before they are reviewed and merged, you can pull the code from my repository instead of this one, using the following: `python -m pip install git+https://github.com/JVemmer/neomodel.git`
As mentioned, the code is not reviewed yet, and therefore no guarantees of working status can be given beyond "it worked for me" and "it passed all the tests".

**Comment by TrietDHA** [comment]
Thank you @JVemmer for the great work! Is there any progress on this pull request? Why doesn't it got accepted yet? I'm really looking forward to using `neomodel` with Neo4j 4.x.

**Comment by davidmoseler** [comment]
I've made a small change to this pull request. If someone is in a hurry and needs to run neomodel with neo4j 4.0, just install it like this:

`python -m pip install git+https://github.com/davidmoseler/neomodel.git`

**Comment by JVemmer** [comment]
> I've made a small change to this pull request. If someone is in a hurry and needs to run neomodel with neo4j 4.0, just install it like this:
> 
> `python -m pip install git+https://github.com/davidmoseler/neomodel.git`

I cherry-picked your commit, so it's included in this pull request :)

**Comment by TrietDHA** [comment]
@JVemmer: is it possible to connect to different databases with your code? So far, `neomodel` only connects to the default active database. Can we somehow provide the database name when creating the connection?

**Comment by JVemmer** [comment]
> @JVemmer: is it possible to connect to different databases with your code? So far, `neomodel` only connects to the default active database. Can we somehow provide the database name when creating the connection?

That's possible in the original code as well. When you define your connection url, just put a forward slash and the name of the database after the port. So instead of e.g. `bolt://neo4j:neo4j@localhost:7687` you use `bolt://neo4j:neo4j@localhost:7687/thisisthedatabasename`

**Comment by TrietDHA** [comment]
> > @JVemmer: is it possible to connect to different databases with your code? So far, `neomodel` only connects to the default active database. Can we somehow provide the database name when creating the connection?
> 
> That's possible in the original code as well. When you define your connection url, just put a forward slash and the name of the database after the port. So instead of e.g. `bolt://neo4j:neo4j@localhost:7687` you use `bolt://neo4j:neo4j@localhost:7687/thisisthedatabasename`

Great! I didn't know that we could do this. I cannot find this anywhere in the documentation.

**Comment by JVemmer** [comment]
> > > @JVemmer: is it possible to connect to different databases with your code? So far, `neomodel` only connects to the default active database. Can we somehow provide the database name when creating the connection?
> > 
> > 
> > That's possible in the original code as well. When you define your connection url, just put a forward slash and the name of the database after the port. So instead of e.g. `bolt://neo4j:neo4j@localhost:7687` you use `bolt://neo4j:neo4j@localhost:7687/thisisthedatabasename`
> 
> Great! I didn't know that we could do this. I cannot find this anywhere in the documentation.

@aanastasiou: It might make sense to update the documentation with this.

**Comment by jherrador** [comment]
Hi @JVemmer,

Do you have news for this PR? Do you know when it will be ready to merge?

Thank you

**Comment by JVemmer** [comment]
> Hi @JVemmer,
> 
> Do you have news for this PR? Do you know when it will be ready to merge?
> 
> Thank you

Hi @jherrador, all the information I have, is here in the pull request. So we are still waiting for @aanastasiou :)

**Comment by aanastasiou** [comment]
@JVemmer and all, I am sorry for the delay, will get back to you shortly about this.

**Comment by TrietDHA** [comment]
Hi @aanastasiou, is there any update for this? I'm looking forward to the new release of Neomodel.

**Comment by whatSocks** [comment]
@tdoan2010 can you rebase your branch?

File: test/test_relationship_models.py [hunk 2]
```diff
@@ -55,8 +55,8 @@ def test_either_connect_with_rel_model():
     # I've tried everything possible to correct this to no avail
     paul = new_rel.start_node()
     tom = new_rel.end_node()
-    assert paul.name == 'Paul'
-    assert tom.name == 'Tom'
+    assert paul.name == 'Tom'
```
Comment (whatSocks) [inline test/test_relationship_models.py:58]: is this correct?

**Comment by whatSocks** [comment]
@JVemmer can you rebase your branch? Thank you for all your work, looking forward to merging soon

**Comment by majiddadashi** [comment]
Hey @whatSocks and @JVemmer! Thank you for your great work. I'm wondering if this PR is going to be merged soon? 

The library is amazing but can't be used right now for those of us that are using Neo4j 4.x. 

I'd also be happy to help to finalize this if you guys point me in the right direction (as I'm new to Neo4j). 

cheers

**Comment by aanastasiou** [comment]
@majiddadashi Please see latest version 4.0.0. Please let us know if you are willing to help with further development. [This list](https://github.com/neo4j-contrib/neomodel/wiki/TODOs-&-Enhancements) of possible work items is fairly up to date, except for the last two issues that were marked as enhancements [here](https://github.com/neo4j-contrib/neomodel/issues/518) and [here](https://github.com/neo4j-contrib/neomodel/issues/505).

**Comment by majiddadashi** [comment]
@aanastasiou Thank you so much! Yes! I'm going to fork the repo and will send PRs as I see issues.

**Pull request closed without merging**